### ISSUE TITLE
Clarification on DDIM Inversion Application in Prior Latent Space

### ISSUE THREAD

**Issue by cbcbcbb** [lixiaowen-xw/DiffuEraser#13]
**Description:**

As a newcomer to diffusion models, I am trying to understand the inference process. In the provided code snippet, random noise is added to the prior latent space as follows:

```python
noise_pre = randn_tensor(shape, device=torch.device(self.device), dtype=prompt_embeds_dtype, generator=generator) 
noise = repeat(noise_pre, "t c h w->(repeat t) c h w", repeat=n_clip)[:real_video_length,...]
```

Subsequently, the VAE encoder processes the pixel values:

```python
with torch.no_grad():
    pixel_values = pixel_values.to(dtype=torch.float16)
    latents = []
    num = 4
    for i in range(0, pixel_values.shape[0], num):
        latents.append(self.vae.encode(pixel_values[i : i + num]).latent_dist.sample())
    latents = torch.cat(latents, dim=0)
latents = latents * self.vae.config.scaling_factor  # [(b f), c1, h, w], c1=4
```

Then, noise is added to the latents using the noise scheduler:

```python
noisy_latents = self.noise_scheduler.add_noise(latents, noise, timesteps)
```

This process appears to involve adding random noise to the prior latent space.

Could you please clarify where the DDIM inversion is applied to the prior latent space within this code?

Thank you for your assistance.

**Comment by lixiaowen-xw** [comment]
Hi, as explained in [](https://huggingface.co/learn/diffusion-course/unit4/2#inversion,) DDIM inversion is about "moving towards higher and higher noise".

**Comment by cbcbcbb** [comment]
Sorry, as I am a beginner, I still haven’t fully grasped the logic behind this. I may have overlooked some details in understanding your project’s implementation, and I would appreciate it if you could point them out. According to my understanding and the implementation from [Hugging Face Diffusion Course](https://huggingface.co/learn/diffusion-course/unit4/2#inversion), DDIM inversion requires:

### Inverted update step (re-arranging the update step to get x(t) (new latents) as a function of x(t-1) (current latents)
```python
latents = (latents - (1 - alpha_t).sqrt() * noise_pred) * (alpha_t_next.sqrt() / alpha_t.sqrt()) + (
    1 - alpha_t_next
).sqrt() * noise_pred
```
This involves a reverse noise addition operation to get the previous latent representation, rather than directly adding noise to the latents obtained from the VAE encoding as done in the project code:
```python
noisy_latents = self.noise_scheduler.add_noise(latents, noise, timesteps)
```
Thank you for your work, and I'll appreciate it  if you can answer my question.
